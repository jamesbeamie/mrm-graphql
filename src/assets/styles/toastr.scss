/* Note that this is toastr v2.1.3, the "latest" version in url has no more maintenance,
* please go to https://cdnjs.com/libraries/toastr.js and pick a certain version you want to use,
* make sure you copy the url from the website since the url may change between versions.
* */
@import './variables.scss';

.toast-message {
  -ms-word-wrap: break-word;
  word-wrap: break-word;
  font-family: "PT Sans Caption", sans-serif;
 }
 .toast-message a,
 .toast-message label {
  color: map-get($color-white, 000);
 }
 .toast-close-button {
  position: relative;
  right: -0.3em;
  top: -0.7em;
  float: right;
  font-size: 20px;
  color: map-get($color-black, 900);
  opacity: 0.8;
  -ms-filter: progid:DXImageTransform.Microsoft.Alpha(Opacity=80);
  filter: alpha(opacity=80);
  line-height: 1;
 }
 .rtl .toast-close-button {
  left: -0.1em;
  float: left;
  right: 0.3em;
 }
 button.toast-close-button {
  padding: 0;
  cursor: pointer;
  background: 0 0;
  border: 0;
  -webkit-appearance: none;
 }
 .toast-top-center {
  top: 0;
  right: 0;
  width: 100%;
 }
 #toast-container {
  position: fixed;
  z-index: 999999;
  pointer-events: none;
 }
 #toast-container * {
  -moz-box-sizing: border-box;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
 }
 #toast-container > div {
  position: relative;
  pointer-events: auto;
  margin: 20px 0 6px;
  padding: 21px 21px 20px 20px;
  -moz-border-radius: 3px;
  -webkit-border-radius: 3px;
  border-radius: 3px;
  background-position: 15px center;
  background-repeat: no-repeat;
  -moz-box-shadow: 0 0 12px map-get($color-black, 100);
  -webkit-box-shadow: 0 0 12px map-get($color-black, 100);
  box-shadow: 0 0 12px map-get($color-black, 100);
  -ms-filter: progid:DXImageTransform.Microsoft.Alpha(Opacity=80);
  filter: alpha(opacity=80);
 }
 #toast-container > div.rtl {
  direction: rtl;
  padding: 15px 50px 15px 15px;
  background-position: right 15px center;
 }
 #toast-container.toast-top-center > div {
  width: 400px;
  margin-left: auto;
  margin-right: auto;
 }
 .toast {
  background-color: map-get($color-black, 900);
 }
 .toast-success {
  background-color: $color-success;
 }
 .toast-error {
  background-color: $color-danger;
  color: $toast-error;
 }
 @media all and (max-width: 240px) {
  #toast-container > div {
    padding: 8px 8px 8px 50px;
    width: 11em;
  }
  #toast-container > div.rtl {
    padding: 8px 50px 8px 8px;
  }
  #toast-container .toast-close-button {
    right: -0.2em;
    top: -0.2em;
  }
  #toast-container .rtl .toast-close-button {
    left: -0.2em;
    right: 0.2em;
  }
 }
 @media all and (min-width: 241px) and (max-width: 480px) {
  #toast-container > div {
    padding: 8px 8px 8px 50px;
    width: 18em;
  }
  #toast-container > div.rtl {
    padding: 8px 50px 8px 8px;
  }
  #toast-container .toast-close-button {
    right: -0.2em;
    top: -0.2em;
  }
  #toast-container .rtl .toast-close-button {
    left: -0.2em;
    right: 0.2em;
  }
 }
 @media all and (min-width: 481px) and (max-width: 768px) {
  #toast-container > div {
    padding: 15px 15px 15px 50px;
    width: 25em;
  }
  #toast-container > div.rtl {
    padding: 15px 50px 15px 15px;
  }
 }