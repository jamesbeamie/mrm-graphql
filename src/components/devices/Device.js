import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import { IconButtons } from '../commons';
import { editIcon, deleteIcon } from '../../utils/images/images';

const formatDate = date => moment(date).format('DD MMM YYYY');

/**
 *
 * Device component
 *
 * @param {Object} device
 *
 * @returns {JSX}
 */
const Device = ({
  handleAction,
  device,
  device: {
    name, deviceType, dateAdded, lastSeen, location, room,
  },
}) => (
  <div className="table__row">
    <span>{name}</span>
    <span>{deviceType}</span>
    <span>{formatDate(dateAdded)}</span>
    <span>{formatDate(lastSeen)}</span>
    <span>{room.name}</span>
    <span>{location}</span>
    <span className="device-table-action-buttons">
      <IconButtons
        btnImage={<img src={editIcon} alt="edit-icon" />}
        className="edit-device-button"
        openModal={() => handleAction('edit', device)}
      />
      <IconButtons
      // delete button
        className="delete-device-button"
        btnImage={<img src={deleteIcon} alt="delete-icon" />}
        openModal={() => handleAction('delete', device)}
      />
    </span>
  </div>
);

Device.propTypes = {
  handleAction: PropTypes.func.isRequired,
  device: PropTypes.shape({
    name: PropTypes.string.isRequired,
    deviceType: PropTypes.string.isRequired,
    dateAdded: PropTypes.string.isRequired,
    lastSeen: PropTypes.string.isRequired,
    room: PropTypes.shape({}).isRequired,
    location: PropTypes.string.isRequired,
  }).isRequired,
};

export default Device;
