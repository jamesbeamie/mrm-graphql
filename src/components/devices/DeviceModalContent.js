import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Input, SelectInput, ActionButtons } from '../commons';

const deleteMessage = formData =>
  (
    <p id="confirm-msg">
      Are you sure you want to delete &quot;{`${formData.name}: ${formData.deviceType}`}
      &quot;? <br />
      This cannot be undone
    </p>
  );

const innerForm = (formData, methods, data) =>
  (
    <Fragment>
      <Input
        id="device-name"
        name="device-name"
        placeholder="Enter Device Name"
        labelName="Device Name"
        labelClass="device-input"
        value={formData.name}
        onChange={methods.handleNameChange}
      />
      <SelectInput
        labelText="Device Type"
        name="device-type"
        id="add-devices-type"
        isValue
        value={formData.deviceType || 'Select Device'}
        options={data.deviceTypes}
        onChange={methods.handleTypeChange}
        wrapperClassName="device-input"
        placeholder="Select Device"
        selectInputClassName="dynamic-input-field default-select"
      />
      <SelectInput
        labelText="Room"
        name="room"
        id="add-devices-rooms"
        value={formData.roomId || 'Select Room'}
        options={data.rooms}
        onChange={methods.handleRoomChange}
        wrapperClassName="device-input"
        placeholder="Select Room"
        selectInputClassName="dynamic-input-field default-select"
      />
    </Fragment>
  );

const DeviceModalContent = (props) => {
  const { formData, methods } = props;
  const {
    modalContent, rooms, deviceTypes, fetching,
  } = props;
  return (
    <form className="amenity-form device-type">
      <div>
        {modalContent !== 'delete' ? innerForm(formData, methods, { deviceTypes, rooms })
          : deleteMessage(formData)}
        <div className="loading-btn-div">
          <ActionButtons
            withCancel
            isLoading={fetching}
            onClickCancel={methods.closeModal}
            actionButtonText={`${modalContent ? ((modalContent !== 'delete') ?
              'SAVE CHANGES' : 'DELETE') : 'SUBMIT'} `}
            onClickSubmit={methods.submit}
          />
        </div>
      </div>
    </form>);
};

DeviceModalContent.propTypes = {
  formData: PropTypes.shape({
    name: PropTypes.string,
    roomId: PropTypes.string,
    deviceType: PropTypes.string,
  }).isRequired,
  methods: PropTypes.shape({
    handleNameChange: PropTypes.func,
    handleRoomChange: PropTypes.func,
    handleTypeChange: PropTypes.func,
    closeModal: PropTypes.func,
    submit: PropTypes.func,
  }).isRequired,
  fetching: PropTypes.bool,
  modalContent: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  rooms: PropTypes.instanceOf(Array).isRequired,
  deviceTypes: PropTypes.instanceOf(Array).isRequired,
};

DeviceModalContent.defaultProps = {
  fetching: false,
  modalContent: undefined,
};

export default DeviceModalContent;
