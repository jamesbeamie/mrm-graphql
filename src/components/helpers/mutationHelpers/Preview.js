import apolloClient from '../../../utils/ApolloClient';
import { ADD_LEVEL_SETUP_MUTATION, DELETE_OFFICE_STRUCTURE } from '../../../graphql/mutations/Preview';
import GET_ALL_LEVELS from '../../../graphql/queries/Levels';

const addLevelSetup = async ({ flattenedData }, client = apolloClient) => {
  await client.mutate({
    mutation: ADD_LEVEL_SETUP_MUTATION,
    name: 'preview',
    variables: {
      flattenedData,
    },
    update: async (store) => {
      // Read the data from our cache for this query.
      const data = store.readQuery({ query: GET_ALL_LEVELS });
      data.allStructures = [...data.allStructures, flattenedData];
      store.writeQuery({
        query: GET_ALL_LEVELS,
        data,
      });
    },
  });
};

export const deleteOfficeStructure = async (structureIds, client = apolloClient) => {
  let data;
  await client.mutate({
    mutation: DELETE_OFFICE_STRUCTURE,
    name: 'delete_office_structure',
    variables: {
      structureIds,
    },
    refetchQueries: [{ query: GET_ALL_LEVELS }],
    update: async (store) => {
      data = store.readQuery({ query: GET_ALL_LEVELS });
      data.allStructures = [...data.allStructures];
      data.allStructures = data.allStructures.filter(structure => structure.state === 'active');
      store.writeQuery({
        query: GET_ALL_LEVELS,
        data,
      });
    },
  });
  return data;
};

export default addLevelSetup;
