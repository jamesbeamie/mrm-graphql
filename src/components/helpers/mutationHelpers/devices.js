/* eslint-disable import/prefer-default-export */
import apolloClient from '../../../utils/ApolloClient';
import { ADD_DEVICE_MUTATION, EDIT_DEVICE_MUTATION, DELETE_DEVICE_MUTATION } from '../../../graphql/mutations/Devices';
import { GET_DEVICES_QUERY } from '../../../graphql/queries/Devices';
import { GET_ROOMS_QUERY } from '../../../graphql/queries/Rooms';

export const addDeviceMutation = async ({
  name, roomId, deviceType, location,
}, client = apolloClient) => {
  await client
    .mutate({
      mutation: ADD_DEVICE_MUTATION,
      name: 'addDeviceMutation',
      variables: {
        name, deviceType, roomId,
      },
      // write and read from cache
      update: async (proxy, { data: { createDevice: { device } } }) => {
        const cachedDevices = proxy.readQuery({
          query: GET_DEVICES_QUERY,
        });
        const deviceList = cachedDevices.allDevices;
        cachedDevices.allDevices = [...deviceList, device];

        proxy.writeQuery({
          query: GET_DEVICES_QUERY,
          data: cachedDevices,
        });

        const cachedRooms = proxy.readQuery({
          query: GET_ROOMS_QUERY,
          variables: {
            page: 1, perPage: 8, location, office: '', roomLabels: '',
          },
        });

        const target = cachedRooms.allRooms.rooms.findIndex(({ id }) => id === roomId);

        const roomsDeviceList = cachedRooms.allRooms.rooms[target].devices;

        cachedRooms.allRooms.rooms[target].devices = [...roomsDeviceList, device];

        proxy.writeQuery({
          query: GET_ROOMS_QUERY,
          variables: {
            location,
            office: '',
            page: 1,
            perPage: 8,
            roomLabels: '',
          },
          data: cachedRooms,
        });
      },
    });
};

export const editDeviceMutation = async (variables, client = apolloClient) => {
  // edit device mutation
  await client.mutate({
    name: 'editDeviceMutation',
    mutation: EDIT_DEVICE_MUTATION,
    variables,
    // write and read from cache
    update: async (proxy, { data: { updateDevice: { device } } }) => {
      const cachedDevices = proxy.readQuery({ query: GET_DEVICES_QUERY });
      const deviceList = cachedDevices.allDevices;
      deviceList.forEach(({ id }, i) => { // update item in cache
        if (id === device.id) { cachedDevices[i] = device; }
      });

      proxy.writeQuery({
        query: GET_DEVICES_QUERY,
        data: cachedDevices,
      });
    },
  });
};

const newDeviceslist = (devicesList, device) => {
  const list = [];
  devicesList.forEach(({ id }, i) => {
    if (id !== device.id) { list.push(devicesList[i]); }
  });
  return list;
};

const updatedCachedRooms = (rooms, target, device) => {
  const targetRoom = target;
  const otherRooms = rooms.filter((room) => {
    const result = room.id !== device.roomId;
    return result;
  });
  targetRoom.devices = newDeviceslist(targetRoom.devices, device);
  otherRooms.push(targetRoom);
  return otherRooms;
};

const cachedRoomslist = (location, proxy) => proxy.readQuery({
  // cachedRooms read from cache
  query: GET_ROOMS_QUERY,
  variables: {
    page: 1,
    perPage: 8,
    location,
    office: '',
    roomLabels: '',
  },
});

const targetRoom = (rooms, device) => rooms.filter((room) => {
  const result = parseInt(room.id, 10) === device.roomId;
  return result;
})[0];

export const deleteDeviceMutation = async (variables, client = apolloClient) => {
  await client
    .mutate({
      mutation: DELETE_DEVICE_MUTATION,
      name: 'deleteDeviceMutation',
      variables, // write and read from cache
      update: async (proxy, { data: { deleteDevice: { device } } }) => {
        try { // write and read from cache
          const cachedDevices = proxy.readQuery({ query: GET_DEVICES_QUERY });
          cachedDevices.allDevices = newDeviceslist(cachedDevices.allDevices, device);
          proxy.writeQuery({ query: GET_DEVICES_QUERY, data: cachedDevices });

          const { location } = variables;
          const cachedRooms = cachedRoomslist(location, proxy);
          const target = targetRoom(cachedRooms.allRooms.rooms, device);

          cachedRooms.allRooms.rooms =
            updatedCachedRooms(cachedRooms.allRooms.rooms, target, device);

          proxy.writeQuery({
            query: GET_ROOMS_QUERY,
            variables: {
              location, office: '', page: 1, perPage: 8, roomLabels: '',
            },
            data: cachedRooms,
          });
        } catch (error) { return null; }
        return null;
      },
    });
};
