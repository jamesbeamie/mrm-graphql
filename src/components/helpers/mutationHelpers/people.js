/* eslint-disable import/prefer-default-export */
import apolloClient from '../../../utils/ApolloClient';
import { INVITE_PERSON_MUTATION, CHANGE_USER_LOCATION_MUTATION } from '../../../graphql/mutations/People';
import { decodeTokenAndGetUserData } from '../../../utils/Cookie';
import { GET_USER_QUERY } from '../../../graphql/queries/People';
import GET_ALL_LEVELS from '../../../graphql/queries/Levels';
import { GET_DEVICES_QUERY } from '../../../graphql/queries/Devices';


const invitePersonMutation = async (email, client = apolloClient) => {
  await client
    .mutate({
      mutation: INVITE_PERSON_MUTATION,
      name: 'invitePersonMutation',
      variables: { email },
    });
};

const changeUserLocation = async (locationId, queryEmail, client = apolloClient) => {
  const { UserInfo: { email: tokenEmail } } = decodeTokenAndGetUserData();
  const email = queryEmail || tokenEmail;

  const { data } = await client
    .mutate({
      mutation: CHANGE_USER_LOCATION_MUTATION,
      name: 'changeUserLocation',
      variables: { email, locationId },
      refetchQueries: [
        { query: GET_USER_QUERY, variables: { email } },
        { query: GET_ALL_LEVELS },
        { query: GET_DEVICES_QUERY },
      ],
    });

  return data;
};

export { invitePersonMutation, changeUserLocation };
