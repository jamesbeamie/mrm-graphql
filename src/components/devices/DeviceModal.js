import React, { Component } from 'react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import notification from '../../utils/notification';
import MrmModal from '../commons/Modal';
import { addDeviceMutation, editDeviceMutation, deleteDeviceMutation } from '../helpers/mutationHelpers/devices';
import DeviceModalContent from './DeviceModalContent';

class DeviceFormModal extends Component {
  state = {
    closeModal: false,
    name: '',
    roomId: '',
    deviceType: '',
    previousFormState: {},
  }

  componentDidUpdate({ openModal: prevOpenModal }) {
    const { openModal, device } = this.props;
    const { closeModal } = this.state;
    if (openModal && !prevOpenModal) {
      this.setDeviceState(device);
    } else if (!openModal && closeModal) {
      this.setCloseModal(false);
      this.updateStateWithStoredData();
    }
  }

  setDeviceState = ({ name, room: { id: roomId }, deviceType }) =>
    this.setState({
      name, roomId, deviceType, closeModal: false,
    });

  setCloseModal = closeModal => this.setState({ closeModal })

  getDeviceTypes = () => ([{
    name: '7"',
  }, {
    name: '8"',
  }, {
    name: '9.7"',
  }, {
    name: '10.1"',
  }])

  updateStateWithStoredData = () => {
    this.setState(({ previousFormState, ...prevState }) =>
      ({ ...prevState, ...previousFormState }));
  };


  handleNameChange = ({ target: { value: name } }) => {
    this.setState({ name });
  };

  handleTypeChange = ({ target: { value: deviceType } }) => {
    this.setState({ deviceType });
  };

  handleRoomChange = ({ target: { value: roomId } }) => {
    this.setState({ roomId });
  };

  closeModal = (e) => {
    if (e) {
      e.preventDefault();
    }
    this.handleFormData(this.props.openModal, this.state.closeModal);
    this.setState({ closeModal: true });
    this.props.closeModal();
  };

  clearForm = wipeHistory => this.setState({
    fetching: false,
    name: '',
    roomId: '',
    deviceType: '',
    ...(wipeHistory && { previousFormState: {} }),
  });

  handleFormData = (openModal, closeModal) => {
    // only clear form if in edit mode, openModal is true only if in edit mode
    if (openModal) {
      this.clearForm();
    } else if (!closeModal) { // store data for modal re-open if in add mode
      this.setState(({ name, roomId, deviceType }) => ({
        previousFormState: {
          name, roomId, deviceType,
        },
      }));
    }
  };

  handleMutation = async () => {
    const { name, roomId, deviceType } = this.state;
    const {
      location: { name: location }, openModal, device, refetch,
    } = this.props;
    const mutations = {
      edit: editDeviceMutation,
      delete: deleteDeviceMutation,
      add: addDeviceMutation,
    };

    const variables = {
      edit: {
        deviceId: device.id,
        name: name || device.name,
        roomId: roomId || device.roomId,
        deviceType: deviceType || device.deviceType,
        location,
      },
      delete: { deviceId: device.id, location },
      add: {
        name, roomId, deviceType, location,
      },
    };

    const mutation = openModal ? mutations[openModal] : mutations.add;
    const variable = openModal ? variables[openModal] : variables.add;

    await mutation(variable);
    return refetch();
  };

  submit = async (e) => {
    e.preventDefault();
    const { name, roomId, deviceType } = this.state;
    const { openModal, closeModal, getRooms } = this.props;
    if (
      (!name || !roomId || !deviceType) && !openModal
    ) {
      return notification(toastr, 'error', 'Please fill all fields')();
    }

    try {
      this.setState({ fetching: true });
      await this.handleMutation();
      this.clearForm(true);
      this.setState({ closeModal: true });
      getRooms();
      closeModal();
      return notification(toastr, 'success', `Device successfully ${openModal ? `${openModal}ed` : 'added'}.`)();
    } catch (error) {
      notification(toastr, 'error', 'Could not complete request.')();
      return this.setState({ fetching: false });
    }
  };

  render() {
    const {
      closeModal: closeModalMethod,
      submit,
      handleNameChange,
      handleRoomChange,
      handleTypeChange,
    } = this;
    const { openModal, rooms } = this.props;
    const {
      closeModal, previousFormState, fetching, ...formData
    } = this.state;
    return (
      <MrmModal
        title={`${!openModal ? 'ADD A NEW' : openModal.toUpperCase()} DEVICE`}
        buttonText="Add a New Device"
        openModal={!!openModal}
        closeModal={closeModal}
        handleCloseRequest={this.closeModal}
        className="add-resource-modal"
      >
        <DeviceModalContent
          formData={formData}
          modalContent={openModal}
          methods={{
            closeModal: closeModalMethod,
            submit,
            handleNameChange,
            handleRoomChange,
            handleTypeChange,
          }}
          fetching={fetching}
          rooms={rooms}
          deviceTypes={this.getDeviceTypes()}
        />
      </MrmModal>
    );
  }
}

DeviceFormModal.propTypes = {
  rooms: PropTypes.instanceOf(Array),
  closeModal: PropTypes.func.isRequired,
  openModal: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
  ]).isRequired,
  refetch: PropTypes.func.isRequired,
  location: PropTypes.shape({
    name: PropTypes.string,
  }),
  device: PropTypes.shape({
    name: PropTypes.string,
    deviceType: PropTypes.string,
    roomId: PropTypes.number,
  }),
};

DeviceFormModal.defaultProps = {
  rooms: [],
  location: undefined,
  device: undefined,
};

export default DeviceFormModal;
