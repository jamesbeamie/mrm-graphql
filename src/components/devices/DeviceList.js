import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import '../../assets/styles/devicelist.scss';
import TableHead from '../helpers/TableHead';
import Device from './Device';
import dummyDevices from '../../fixtures/devices';
import { getAllDevices, getAllRooms } from '../helpers/QueriesHelpers';
import Overlay from '../commons/Overlay';
import ErrorIcon from '../commons/ErrorIcon';
import Pagination from '../commons/Pagination';
import paginate from '../helpers/FrontendPagination';
import DeviceModal from './DeviceModal';

/**
 * Device List Component
 *
 * @returns {JSX}
 */
class DeviceList extends Component {
  state = {
    fetching: true,
    devices: [],
    rooms: [],
    perPage: 5,
    currentPage: 1,
    openModal: false,
    device: {},
  }

  async componentDidMount() {
    this.getData();
  }

  componentDidUpdate(prevProps, { currentPage, perPage }) {
    if (
      currentPage !== this.state.currentPage
      || perPage !== this.state.perPage
    ) {
      this.updatePageDetails();
    }
    if (prevProps.userLocationChanged !== this.props.userLocationChanged) {
      this.toggleFetching();
      setTimeout(() => { this.getData(); }, 1000);
      this.toggleFetching();
    }
  }

  getData = async () => {
    const { location: { name: location, id } } = this.props;
    const devices = await getAllDevices();
    const { allRooms: { rooms: allRooms } } = await getAllRooms(location);
    const rooms = allRooms.filter(({ locationId }) => locationId === Number(id));
    this.setState({
      devices: [...devices].reverse(),
      rooms,
    });
    this.updatePageDetails();
  }

  toggleFetching() { this.setState({ fetching: !this.state.fetching }); }

  handleAction = (action, device) => {
    this.setState({
      openModal: action,
      device,
    });
  }

  deviceComponents = deviceArray =>
    deviceArray.map(device =>
      <Device handleAction={this.handleAction} device={device} key={device.name + device.id} />);


  handlePageChange = (perPageData = 5, currentPageData) => {
    this.setState({
      perPage: Number(perPageData),
      currentPage: Number(currentPageData),
    });
  }

  closeModal = () => this.setState({
    openModal: false,
    modalIsClosed: true,
    device: {},
  });

  updatePageDetails = () => {
    const {
      devices, currentPage, fetching, perPage,
    } = this.state;
    const {
      pageContent: paginatedDeviceList, hasNext, hasPrevious, pages,
    } = paginate(devices,
      { currentPage, perPage },
    );

    this.setState({
      paginatedDeviceList,
      hasNext,
      hasPrevious,
      pages,
      ...(fetching && { fetching: false }),
    });
  }
  render() {
    const {
      device,
      fetching,
      rooms,
      devices,
      currentPage,
      modalIsClosed,
      paginatedDeviceList,
      pages,
      hasNext,
      hasPrevious,
      perPage,
      openModal,
    } = this.state;
    const { location, getRooms } = this.props;

    const displayPaginator = !fetching
      && !(devices.length < 5 && currentPage === 1);

    return (
      <div className="settings-devices-list overlay-container">
        {fetching && <Overlay />}
        <div className="settings-devices-control" />
        <div className="add-new-resource">
          <DeviceModal
            openModal={openModal}
            closeModal={this.closeModal}
            location={location}
            rooms={rooms}
            refetch={this.getData}
            device={device}
            getRooms={getRooms}
            modalIsClosed={modalIsClosed}
          />
        </div>
        {
          !fetching && !devices[0]
            ? (
              <ErrorIcon message="No Device found" />
            ) : (
              <Fragment>
                <div className="table device-table">
                  <TableHead titles={['Name', 'Type', 'Date Added', 'Last Seen', 'Room', 'Location']} />
                  <div className="table__body">{this.deviceComponents((
                    fetching ? dummyDevices : paginatedDeviceList
                  ))}
                  </div>
                </div>
                {
                  displayPaginator &&
                  <div>
                    <Pagination
                      totalPages={pages}
                      hasNext={hasNext}
                      hasPrevious={hasPrevious}
                      handleData={this.handlePageChange}
                      isFetching={fetching}
                      perPage={perPage}
                      currentPage={currentPage}
                    />
                  </div>
                }
              </Fragment>
            )
        }
      </div>
    );
  }
}


DeviceList.propTypes = {
  userLocationChanged: PropTypes.bool,
  location: PropTypes.shape({
    name: PropTypes.string,
  }),
  getRooms: PropTypes.func.isRequired,
};

DeviceList.defaultProps = {
  userLocationChanged: PropTypes.bool,
  location: undefined,
};


export default DeviceList;
