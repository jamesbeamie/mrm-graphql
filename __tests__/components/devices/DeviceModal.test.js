import React from 'react';
import { shallow } from 'enzyme';
import toastr from 'toastr';
import DeviceModal from '../../../src/components/devices/DeviceModal';
import { addDeviceMutation } from '../../../src/components/helpers/mutationHelpers/devices';


jest.mock('../../../src/components/helpers/mutationHelpers/devices');
jest.mock('toastr');

describe('Tests for DeviceFormModal', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = shallow(<DeviceModal
      location={{
      name: 'Lagos',
    }}
      refetch={jest.fn()}
      closeModal={jest.fn()}
      getRooms={jest.fn()}
      device={{ room: {} }}
    />);
  });

  it('Should render correctly from memory', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should handle submit correctly with correct data', async () => {
    addDeviceMutation.mockResolvedValue({});
    wrapper.setState({
      name: 'Galaxy S9',
      roomId: '32',
      deviceType: 'Tablet',
    });
    await wrapper.instance().submit({ preventDefault: jest.fn() });
    wrapper.setProps({ openModal: 'edit' });
    await wrapper.instance().submit({ preventDefault: jest.fn() });
    expect(addDeviceMutation).toHaveBeenCalled();
    expect(toastr.success).toHaveBeenCalled();
    wrapper.setProps({ openModal: undefined });
  });

  it('Should handle server errors', async () => {
    wrapper.setState({
      name: 'Galaxy S9',
      roomId: '32',
      deviceType: 'Tablet',
    });
    addDeviceMutation.mockRejectedValue({});
    await wrapper.instance().submit({ preventDefault: jest.fn() });
    expect(toastr.error).toHaveBeenCalledWith('Could not complete request.');
  });


  it('Should not submit with incomplete data', () => {
    wrapper.setState({
      deviceType: undefined,
    });
    wrapper.instance().submit({ preventDefault: jest.fn() });
    expect(toastr.error).toHaveBeenCalledWith('Please fill all fields');
  });

  it('Should call setter methods', () => {
    const getInputObject = value => ({
      target: {
        value,
      },
    });

    wrapper.instance().handleNameChange(getInputObject('Tab'));
    expect(wrapper.state('name')).toBe('Tab');

    wrapper.instance().handleTypeChange(getInputObject('Tablet'));
    expect(wrapper.state('deviceType')).toBe('Tablet');

    wrapper.instance().handleRoomChange(getInputObject('123'));
    expect(wrapper.state('roomId')).toBe('123');
  });

  it('Should use device if provided', () => {
    wrapper.setProps({
      openModal: 'edit',
      device: {
        name: 'Galaxy S10',
        room: {
          id: '31',
        },
        deviceType: 'Phone',
      },
    });
    wrapper.instance().componentDidUpdate({ openModal: false });
    expect(wrapper.state('roomId')).toBe('31');
  });

  describe('Should close modal', () => {
    it('Should close modal and not save if in edit or delete mode', () => {
      wrapper.instance().closeModal();
      expect(wrapper.state('closeModal')).toBe(true);
    });

    it('Should close modal and save if in add mode', () => {
      wrapper.setProps({ openModal: undefined });
      wrapper.setState({ closeModal: false });
      wrapper.instance().closeModal({ preventDefault: jest.fn() });
      expect(wrapper.state('previousFormState')).toEqual({
        name: '',
        roomId: '',
        deviceType: '',
      });
    });
  });

  it('handleFormData should do nothing if no criteria is met', () => {
    wrapper.instance().handleFormData(false, true);
    expect(wrapper.state('closeModal')).toEqual(false);
  });
});
