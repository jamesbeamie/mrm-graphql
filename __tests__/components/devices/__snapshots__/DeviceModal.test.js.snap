// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for DeviceFormModal Should render correctly from memory 1`] = `
<MrmModal
  buttonText="Add a New Device"
  className="add-resource-modal"
  closeModal={false}
  handleCloseRequest={[Function]}
  modalButtonClassName="button"
  openModal={false}
  title="ADD A NEW DEVICE"
>
  <DeviceModalContent
    deviceTypes={
      Array [
        Object {
          "name": "7\\"",
        },
        Object {
          "name": "8\\"",
        },
        Object {
          "name": "9.7\\"",
        },
        Object {
          "name": "10.1\\"",
        },
      ]
    }
    fetching={false}
    formData={
      Object {
        "deviceType": "",
        "name": "",
        "roomId": "",
      }
    }
    methods={
      Object {
        "closeModal": [Function],
        "handleNameChange": [Function],
        "handleRoomChange": [Function],
        "handleTypeChange": [Function],
        "submit": [Function],
      }
    }
    rooms={Array []}
  />
</MrmModal>
`;
